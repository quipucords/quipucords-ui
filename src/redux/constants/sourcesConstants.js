const ADD_SOURCE = 'ADD_SOURCE';
const DELETE_SOURCE = 'DELETE_SOURCE';
const DELETE_SOURCES = 'DELETE_SOURCES';
const GET_SCANS_SOURCES = 'GET_SCANS_SOURCES';
const GET_SOURCES = 'GET_SOURCES';
const UPDATE_SOURCES = 'UPDATE_SOURCES';
const UPDATE_SOURCE = 'UPDATE_SOURCE';
const VALID_SOURCE_WIZARD_STEPONE = 'VALID_SOURCE_WIZARD_STEPONE';
const VALID_SOURCE_WIZARD_STEPTWO = 'VALID_SOURCE_WIZARD_STEPTWO';
const INVALID_SOURCE_WIZARD_STEPTWO = 'INVALID_SOURCE_WIZARD_STEPTWO';
const CREATE_SOURCE_SHOW = 'CREATE_SOURCE_SHOW';
const EDIT_SOURCE_SHOW = 'EDIT_SOURCE_SHOW';
const UPDATE_SOURCE_HIDE = 'UPDATE_SOURCE_HIDE';
const SELECT_SOURCE = 'SELECT_SOURCE';
const DESELECT_SOURCE = 'DESELECT_SOURCE';
const EXPANDED_SOURCE = 'EXPANDED_SOURCE';
const NOT_EXPANDED_SOURCE = 'NOT_EXPANDED_SOURCE';

const sourcesTypes = {
  ADD_SOURCE,
  DELETE_SOURCE,
  DELETE_SOURCES,
  GET_SCANS_SOURCES,
  GET_SOURCES,
  UPDATE_SOURCES,
  UPDATE_SOURCE,
  VALID_SOURCE_WIZARD_STEPONE,
  VALID_SOURCE_WIZARD_STEPTWO,
  INVALID_SOURCE_WIZARD_STEPTWO,
  CREATE_SOURCE_SHOW,
  EDIT_SOURCE_SHOW,
  UPDATE_SOURCE_HIDE,
  SELECT_SOURCE,
  DESELECT_SOURCE,
  EXPANDED_SOURCE,
  NOT_EXPANDED_SOURCE
};

export {
  sourcesTypes as default,
  sourcesTypes,
  ADD_SOURCE,
  DELETE_SOURCE,
  DELETE_SOURCES,
  GET_SCANS_SOURCES,
  GET_SOURCES,
  UPDATE_SOURCES,
  UPDATE_SOURCE,
  VALID_SOURCE_WIZARD_STEPONE,
  VALID_SOURCE_WIZARD_STEPTWO,
  INVALID_SOURCE_WIZARD_STEPTWO,
  CREATE_SOURCE_SHOW,
  EDIT_SOURCE_SHOW,
  UPDATE_SOURCE_HIDE,
  SELECT_SOURCE,
  DESELECT_SOURCE,
  EXPANDED_SOURCE,
  NOT_EXPANDED_SOURCE
};
