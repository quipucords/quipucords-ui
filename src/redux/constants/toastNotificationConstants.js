const TOAST_ADD = 'TOAST_ADD';
const TOAST_REMOVE = 'TOAST_REMOVE';
const TOAST_PAUSE = 'TOAST_PAUSE';
const TOAST_RESUME = 'TOAST_RESUME';

const toastNotificationTypes = {
  TOAST_ADD,
  TOAST_REMOVE,
  TOAST_PAUSE,
  TOAST_RESUME
};

export {
  toastNotificationTypes as default,
  toastNotificationTypes,
  TOAST_ADD,
  TOAST_REMOVE,
  TOAST_PAUSE,
  TOAST_RESUME
};
