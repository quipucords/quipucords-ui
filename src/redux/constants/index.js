import { aboutModalTypes } from './aboutModalConstants';
import { confirmationModalTypes } from './confirmationModalConstants';
import { credentialsTypes } from './credentialsConstants';
import { factsTypes } from './factsConstants';
import { reportsTypes } from './reportsConstants';
import { scansTypes } from './scansConstants';
import { sourcesTypes } from './sourcesConstants';
import { statusTypes } from './statusConstants';
import { toastNotificationTypes } from './toastNotificationConstants';
import { userTypes } from './userConstants';
import { viewTypes } from './viewConstants';

const reduxTypes = {
  aboutModal: aboutModalTypes,
  confirmationModal: confirmationModalTypes,
  credentials: credentialsTypes,
  facts: factsTypes,
  reports: reportsTypes,
  scans: scansTypes,
  sources: sourcesTypes,
  status: statusTypes,
  toastNotifications: toastNotificationTypes,
  user: userTypes,
  view: viewTypes
};

export {
  reduxTypes as default,
  reduxTypes,
  aboutModalTypes,
  confirmationModalTypes,
  credentialsTypes,
  factsTypes,
  reportsTypes,
  scansTypes,
  sourcesTypes,
  statusTypes,
  toastNotificationTypes,
  userTypes,
  viewTypes
};
