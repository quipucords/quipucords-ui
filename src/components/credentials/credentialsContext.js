import React, { useState } from 'react';
import { useShallowCompareEffect } from 'react-use';
import { AlertVariant, List, ListItem } from '@patternfly/react-core';
import { ContextIcon, ContextIconVariant } from '../contextIcon/contextIcon';
import { reduxActions, reduxTypes, storeHooks } from '../../redux';
import { useConfirmation } from '../../hooks/useConfirmation';
import { useOnUpdateCredential } from '../createCredentialDialog/createCredentialDialogContext';
import { useView } from '../view/viewContext';
import { API_QUERY_SORT_TYPES, API_QUERY_TYPES, apiTypes } from '../../constants/apiConstants';
import { translate } from '../i18n/i18n';

/**
 * State context identifier
 *
 * @type {string}
 */
const VIEW_ID = 'credentials';

/**
 * Charge initial view query
 *
 * @type {{'[API_QUERY_TYPES.ORDERING]': string, '[API_QUERY_TYPES.PAGE]': number, '[API_QUERY_TYPES.PAGE_SIZE]': number}}
 */
const INITIAL_QUERY = {
  [API_QUERY_TYPES.ORDERING]: API_QUERY_SORT_TYPES.NAME,
  [API_QUERY_TYPES.PAGE]: 1,
  [API_QUERY_TYPES.PAGE_SIZE]: 10
};

/**
 * Credential action, onDelete.
 *
 * @param {object} options
 * @param {Function} options.deleteCredentials
 * @param {Function} options.t
 * @param {Function} options.useConfirmation
 * @param {Function} options.useDispatch
 * @param {Function} options.useSelectorsResponse
 * @param {Function} options.useView
 * @returns {(function(*): void)|*}
 */
const useOnDelete = ({
  deleteCredentials = reduxActions.credentials.deleteCredential,
  t = translate,
  useConfirmation: useAliasConfirmation = useConfirmation,
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useSelectorsResponse: useAliasSelectorsResponse = storeHooks.reactRedux.useSelectorsResponse,
  useView: useAliasView = useView
} = {}) => {
  const { viewId } = useAliasView();
  const onConfirmation = useAliasConfirmation();
  const [credentialsToDelete, setCredentialsToDelete] = useState([]);
  const dispatch = useAliasDispatch();
  const { data, error, fulfilled } = useAliasSelectorsResponse(({ credentials }) => credentials?.deleted);
  const { errorMessage } = data?.[0] || {};

  useShallowCompareEffect(() => {
    if (credentialsToDelete.length) {
      const credentialIds = credentialsToDelete.map(cred => cred[apiTypes.API_RESPONSE_CREDENTIAL_ID]);
      deleteCredentials(credentialIds)(dispatch);
    }
  }, [credentialsToDelete, deleteCredentials, dispatch]);

  useShallowCompareEffect(() => {
    if ((fulfilled && credentialsToDelete.length) || (error && credentialsToDelete.length)) {
      const credentialNames = credentialsToDelete.map(cred => cred[apiTypes.API_RESPONSE_CREDENTIAL_NAME]);

      dispatch([
        {
          type: reduxTypes.toastNotifications.TOAST_ADD,
          alertType: (error && AlertVariant.danger) || AlertVariant.success,
          header: t('toast-notifications.title', {
            context: [(error && 'error') || 'deleted-credential'],
            count: credentialNames.length
          }),
          message: t('toast-notifications.description', {
            context: ['deleted-credential', error && 'error'],
            name: credentialNames[0],
            count: credentialNames.length,
            message: errorMessage
          })
        },
        {
          type: reduxTypes.credentials.DESELECT_CREDENTIAL,
          item: credentialsToDelete
        },
        {
          type: reduxTypes.credentials.RESET_ACTIONS
        },
        {
          type: reduxTypes.view.UPDATE_VIEW,
          viewId
        }
      ]);

      setCredentialsToDelete(() => []);
    }
  }, [credentialsToDelete, error, fulfilled, dispatch, errorMessage, t]);

  /**
   * Confirmation, and state, for deleting a single or multiple credentials
   *
   * @param {Array|string} credentials
   */
  return credentials => {
    const updatedCredentials = (Array.isArray(credentials) && credentials) || [credentials];

    onConfirmation({
      title: t('form-dialog.confirmation', {
        context: ['title', 'delete-credential'],
        count: updatedCredentials.length
      }),
      heading: t(
        'form-dialog.confirmation',
        {
          context: ['heading', 'delete-credential'],
          count: updatedCredentials.length,
          name: updatedCredentials?.[0]?.[apiTypes.API_RESPONSE_CREDENTIAL_NAME]
        },
        [<strong />]
      ),
      body:
        (updatedCredentials.length > 1 && (
          <List className="quipucords-list__overflow-scroll" isPlain>
            {updatedCredentials.map(
              ({
                [apiTypes.API_RESPONSE_CREDENTIAL_NAME]: name,
                [apiTypes.API_RESPONSE_CREDENTIAL_CRED_TYPE]: credType
              }) => (
                <ListItem key={name} icon={<ContextIcon symbol={ContextIconVariant[credType]} />}>
                  {name}
                </ListItem>
              )
            )}
          </List>
        )) ||
        undefined,
      confirmButtonText: t('form-dialog.label', {
        context: ['delete']
      }),
      onConfirm: () => setCredentialsToDelete(() => updatedCredentials)
    });
  };
};

/**
 * On edit a credential, show modal.
 *
 * @param {object} options
 * @param {Function} options.useOnUpdateCredential
 * @returns {Function}
 */
const useOnEdit = ({ useOnUpdateCredential: useAliasOnUpdateCredential = useOnUpdateCredential } = {}) => {
  const { onEdit } = useAliasOnUpdateCredential();
  return onEdit;
};

/**
 * On expand a row facet.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {Function}
 */
const useOnExpand = ({ useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  return ({ isExpanded, cellIndex, data }) => {
    dispatch({
      type: isExpanded ? reduxTypes.credentials.EXPANDED_CREDENTIAL : reduxTypes.credentials.NOT_EXPANDED_CREDENTIAL,
      viewType: reduxTypes.view.CREDENTIALS_VIEW,
      item: data.item,
      cellIndex
    });
  };
};

/**
 * On select a row.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {Function}
 */
const useOnSelect = ({ useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  return ({ isSelected, data }) => {
    dispatch({
      type: isSelected ? reduxTypes.credentials.SELECT_CREDENTIAL : reduxTypes.credentials.DESELECT_CREDENTIAL,
      viewType: reduxTypes.view.CREDENTIALS_VIEW,
      item: data.item
    });
  };
};

/**
 * Use credentials' response
 *
 * @param {object} options
 * @param {Function} options.useSelectors
 * @param {Function} options.useSelectorsResponse
 * @returns {{date: *, totalResults: (*|number), data: *[], pending: boolean, hasData: boolean, errorMessage: null,
 *     fulfilled: boolean, selectedRows: *, expandedRows: *, error: boolean}}
 */
const useCredentials = ({
  useSelectors: useAliasSelectors = storeHooks.reactRedux.useSelectors,
  useSelectorsResponse: useAliasSelectorsResponse = storeHooks.reactRedux.useSelectorsResponse
} = {}) => {
  const [selectedRows, expandedRows] = useAliasSelectors([
    ({ credentials }) => credentials?.selected,
    ({ credentials }) => credentials?.expanded
  ]);
  const {
    data: responseData,
    error,
    fulfilled,
    message: errorMessage,
    pending,
    responses = {}
  } = useAliasSelectorsResponse({ id: 'view', selector: ({ credentials }) => credentials?.view });

  const [{ date } = {}] = responses?.list || [];
  const {
    [apiTypes.API_RESPONSE_CREDENTIALS_COUNT]: totalResults,
    [apiTypes.API_RESPONSE_CREDENTIALS_RESULTS]: data = []
  } = responseData?.view || {};

  return {
    pending,
    error,
    errorMessage,
    fulfilled,
    data,
    date,
    hasData: fulfilled === true && totalResults > 0,
    selectedRows,
    expandedRows,
    totalResults: totalResults || 0
  };
};

/**
 * Get credentials
 *
 * @param {object} options
 * @param {Function} options.getCredentials
 * @param {Function} options.useCredentials
 * @param {Function} options.useDispatch
 * @param {Function} options.useSelectors
 * @param {Function} options.useView
 * @returns {{date: *, data: *[], pending: boolean, errorMessage: null, fulfilled: boolean, selectedRows: *,
 *     expandedRows: *, error: boolean}}
 */
const useGetCredentials = ({
  getCredentials = reduxActions.credentials.getCredentials,
  useCredentials: useAliasCredentials = useCredentials,
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useSelectors: useAliasSelectors = storeHooks.reactRedux.useSelectors,
  useView: useAliasView = useView
} = {}) => {
  const { query, viewId } = useAliasView();
  const dispatch = useAliasDispatch();
  const [refreshUpdate] = useAliasSelectors([({ view }) => view.update?.[viewId]]);
  const response = useAliasCredentials();

  useShallowCompareEffect(() => {
    getCredentials(null, query)(dispatch);
  }, [dispatch, getCredentials, query, refreshUpdate]);

  return response;
};

const context = {
  VIEW_ID,
  INITIAL_QUERY,
  useCredentials,
  useGetCredentials,
  useOnDelete,
  useOnEdit,
  useOnExpand,
  useOnSelect
};

export {
  context as default,
  context,
  VIEW_ID,
  INITIAL_QUERY,
  useCredentials,
  useGetCredentials,
  useOnDelete,
  useOnEdit,
  useOnExpand,
  useOnSelect
};
