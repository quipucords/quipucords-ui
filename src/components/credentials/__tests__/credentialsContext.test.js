import { context, useCredentials, useGetCredentials, useOnDelete } from '../credentialsContext';
import { apiTypes } from '../../../constants/apiConstants';

describe('CredentialsContext', () => {
  it('should return specific properties', () => {
    expect(context).toMatchSnapshot('specific properties');
  });

  it('should handle deleting a credential with a confirmation', async () => {
    const mockConfirmation = jest.fn();
    const mockCredential = {
      [apiTypes.API_RESPONSE_CREDENTIAL_ID]: 'lorem ipsum base id',
      [apiTypes.API_RESPONSE_CREDENTIAL_NAME]: 'lorem ipsum name'
    };

    const { result } = await shallowHook(() =>
      useOnDelete({
        useConfirmation: () => mockConfirmation
      })
    );

    result(mockCredential);

    expect(mockConfirmation.mock.calls).toMatchSnapshot('dispatch onDelete');
    mockConfirmation.mockClear();
  });

  it('should apply a hook for retrieving data from multiple selectors', () => {
    const { result: errorResponse } = shallowHook(() =>
      useCredentials({
        useSelectorsResponse: () => ({ error: true, message: 'Lorem ipsum' })
      })
    );

    const { result: pendingResponse } = shallowHook(() =>
      useCredentials({
        useSelectorsResponse: () => ({ pending: true })
      })
    );

    const { result: fulfilledResponse } = shallowHook(() =>
      useCredentials({
        useSelectorsResponse: () => ({ fulfilled: true, data: { view: { results: ['dolor', 'sit'] } } })
      })
    );

    const { result: mockStoreSuccessResponse } = shallowHook(() => useCredentials(), {
      state: {
        view: {
          update: {}
        },
        credentials: {
          expanded: {},
          selected: {},
          view: {
            fulfilled: true,
            data: {
              results: ['lorem', 'ipsum']
            }
          }
        }
      }
    });

    expect({ errorResponse, fulfilledResponse, pendingResponse, mockStoreSuccessResponse }).toMatchSnapshot(
      'selector responses'
    );
  });

  it('should apply a hook for returning a get response', () => {
    const { result: errorResponse } = shallowHook(() =>
      useGetCredentials({
        useCredentials: () => ({ error: true, message: 'Lorem ipsum' })
      })
    );

    const { result: pendingResponse } = shallowHook(() =>
      useGetCredentials({
        useCredentials: () => ({ pending: true })
      })
    );

    const { result: fulfilledResponse } = shallowHook(() =>
      useGetCredentials({
        useCredentials: () => ({ fulfilled: true, data: { view: { results: ['dolor', 'sit'] } } })
      })
    );

    const { result: mockStoreSuccessResponse } = shallowHook(() => useGetCredentials(), {
      state: {
        view: {
          update: {}
        },
        credentials: {
          expanded: {},
          selected: {},
          view: {
            fulfilled: true,
            data: {
              results: ['lorem', 'ipsum']
            }
          }
        }
      }
    });

    expect({ errorResponse, fulfilledResponse, pendingResponse, mockStoreSuccessResponse }).toMatchSnapshot(
      'get responses'
    );
  });
});
