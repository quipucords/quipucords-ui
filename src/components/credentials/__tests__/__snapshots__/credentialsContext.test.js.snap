// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CredentialsContext should apply a hook for retrieving data from multiple selectors: selector responses 1`] = `
Object {
  "errorResponse": Object {
    "data": Array [],
    "date": undefined,
    "error": true,
    "errorMessage": "Lorem ipsum",
    "expandedRows": undefined,
    "fulfilled": undefined,
    "hasData": false,
    "pending": undefined,
    "selectedRows": undefined,
    "totalResults": 0,
  },
  "fulfilledResponse": Object {
    "data": Array [
      "dolor",
      "sit",
    ],
    "date": undefined,
    "error": undefined,
    "errorMessage": undefined,
    "expandedRows": undefined,
    "fulfilled": true,
    "hasData": false,
    "pending": undefined,
    "selectedRows": undefined,
    "totalResults": 0,
  },
  "mockStoreSuccessResponse": Object {
    "data": Array [
      "lorem",
      "ipsum",
    ],
    "date": undefined,
    "error": false,
    "errorMessage": null,
    "expandedRows": Object {},
    "fulfilled": true,
    "hasData": false,
    "pending": false,
    "selectedRows": Object {},
    "totalResults": 0,
  },
  "pendingResponse": Object {
    "data": Array [],
    "date": undefined,
    "error": undefined,
    "errorMessage": undefined,
    "expandedRows": undefined,
    "fulfilled": undefined,
    "hasData": false,
    "pending": true,
    "selectedRows": undefined,
    "totalResults": 0,
  },
}
`;

exports[`CredentialsContext should apply a hook for returning a get response: get responses 1`] = `
Object {
  "errorResponse": Object {
    "error": true,
    "message": "Lorem ipsum",
  },
  "fulfilledResponse": Object {
    "data": Object {
      "view": Object {
        "results": Array [
          "dolor",
          "sit",
        ],
      },
    },
    "fulfilled": true,
  },
  "mockStoreSuccessResponse": Object {
    "data": Array [
      "lorem",
      "ipsum",
    ],
    "date": undefined,
    "error": false,
    "errorMessage": null,
    "expandedRows": Object {},
    "fulfilled": true,
    "hasData": false,
    "pending": false,
    "selectedRows": Object {},
    "totalResults": 0,
  },
  "pendingResponse": Object {
    "pending": true,
  },
}
`;

exports[`CredentialsContext should handle deleting a credential with a confirmation: dispatch onDelete 1`] = `
Array [
  Array [
    Object {
      "body": undefined,
      "confirmButtonText": "t(form-dialog.label, {\\"context\\":\\"delete\\"})",
      "heading": "t(form-dialog.confirmation_heading, {\\"context\\":\\"delete-credential\\",\\"count\\":1,\\"name\\":\\"lorem ipsum name\\"}, [object Object])",
      "onConfirm": [Function],
      "title": "t(form-dialog.confirmation_title, {\\"context\\":\\"delete-credential\\",\\"count\\":1})",
    },
  ],
]
`;

exports[`CredentialsContext should return specific properties: specific properties 1`] = `
Object {
  "INITIAL_QUERY": Object {
    "ordering": "name",
    "page": 1,
    "page_size": 10,
  },
  "VIEW_ID": "credentials",
  "useCredentials": [Function],
  "useGetCredentials": [Function],
  "useOnDelete": [Function],
  "useOnEdit": [Function],
  "useOnExpand": [Function],
  "useOnSelect": [Function],
}
`;
